### PR TITLE
Updates to the README

### PR DESCRIPTION
**Pull request by richvdh**
Attempt to drag the README into 2016.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -66,26 +67,16 @@ hosted by someone else (e.g. matrix.org) - there is no single point of control
 or mandatory service provider in Matrix, unlike WhatsApp, Facebook, Hangouts,
 etc.
 
-Synapse ships with two basic demo Matrix clients: webclient (a basic group chat
-web client demo implemented in AngularJS) and cmdclient (a basic Python
-command line utility which lets you easily see what the JSON APIs are up to).
-
-Meanwhile, iOS and Android SDKs and clients are available from:
-
-- https://github.com/matrix-org/matrix-ios-sdk
```
Comment (ara4n) [inline README.rst]: It feels like a shame to remove the links to the official SDKs here.  I know we also link to the huge project page, but saying "hey, the official SDKs for JS, iOS & Android are here - please go get started!" feels worthwhile.

File: README.rst [hunk 2]
```diff
@@ -66,26 +67,16 @@ hosted by someone else (e.g. matrix.org) - there is no single point of control
...
+https://matrix.org/docs/projects/try-matrix-now), run a homeserver, take a look
+at the Matrix spec at https://matrix.org/docs/spec and API docs at
 https://matrix.org/docs/api, experiment with the APIs and the demo clients, and
-report any bugs via https://matrix.org/jira.
+report any bugs via github.
 
 Thanks for using Matrix!
 
-[1] End-to-end encryption is currently in development - see https://matrix.org/git/olm
+[1] End-to-end encryption is currently in beta.
```
Comment (ara4n) [inline README.rst]: This is a great place surely to link to info on e2e, which will be the next question of anyone reading it.  Surely we should link to it, or at least to the beta blog post or something?

File: README.rst [hunk 3]
```diff
@@ -151,38 +142,65 @@ This installs synapse, along with the libraries it uses, into a virtual
...
         --report-stats=[yes|no]
 
-...substituting your host and domain name as appropriate.
-
-This will generate you a config file that you can then customise, but it will
+... substituting an appropriate value for ``--server-name``. The server name
+determines the "domain" part of user-ids for users on your server: these will
+all be of the format ``@user:my.domain.name``. It also determines how other
+matrix servers will reach yours for `Federation`_. For a test configuration,
+set this to the hostname of your server. For a more production-ready setup, you
```
Comment (ara4n) [inline README.rst]: I think we should warn people to get the right name from the get go, given they can't rename servers.

File: README.rst [hunk 4]
```diff
@@ -454,140 +478,238 @@ versions of synapse.
...
----------------------------------
+If you want to set up a reverse-proxy on the federation port despite these
+caveats, you will need to do the following:
+
+* In ``homeserver.yaml``, set ``tls_certificate_path`` to the path to the SSL
+  certificate file used by your reverse-proxy, and set ``no_tls`` to ``True``.
+  (``tls_private_key_path`` will be ignored if ``no_tls`` is ``True``.)
+
+* In your reverse-proxy configuration, if there are other virtual hosts on the
+  same port, make sure that Synapse is the default.
```
Comment (uhoreg) [inline README.rst]: Maybe make a note that Synapse only needs `/_matrix/*` forwarded, so the proxy can forward other paths in the default virtual host to another server?  Not sure of the best wording.

File: README.rst [hunk 5]
```diff
@@ -151,38 +142,65 @@ This installs synapse, along with the libraries it uses, into a virtual
...
+doing so may require more advanced setup - see `Setting up Federation`_.
+
+This command will generate you a config file that you can then customise, but it will
 also generate a set of keys for you. These keys will allow your Home Server to
 identify itself to other Home Servers, so don't lose or delete them. It would be
 wise to back them up somewhere safe. If, for whatever reason, you do need to
 change your Home Server's keys, you may find that other Home Servers have the
 old key cached. If you update the signing key, you should change the name of the
-key in the <server name>.signing.key file (the second word) to something different.
+key in the ``<server name>.signing.key`` file (the second word) to something different.
```
Comment (ara4n) [inline README.rst]: this obviously begs the question as to how keys & perspectives are validated. i suspect we should update it with a quick explanation of wherever the key validation stuff is at.

File: README.rst [hunk 6]
```diff
@@ -151,38 +142,65 @@ This installs synapse, along with the libraries it uses, into a virtual
...
+
+The default configuration exposes two TCP ports: 8008 and 8448. Port 8008 is
```
Comment (ara4n) [inline README.rst]: surely we should spell out whether these are for federation, CS, or both, or what

File: README.rst [hunk 7]
```diff
@@ -205,11 +234,54 @@ run (e.g. ``~/.synapse``), and::
     source ./bin/activate
     synctl start
 
+
+Connecting to Synapse from a client
+===================================
+
+The easiest way to try out your new Synapse installation is by connecting to it
+from a web client. We recommend the one at http://riot.im/app. You will need to
```
Comment (ara4n) [inline README.rst]: as matrix.org, we really can't specifically recommend Riot.  i suggest "The easiest option is probably..." or "One option is..."

File: README.rst [hunk 8]
```diff
@@ -454,140 +478,238 @@ versions of synapse.
...
+If you've already generated the config file, you need to edit the ``server_name``
+in your ``homeserver.yaml`` file. If you've already started Synapse and a
+database has been created, you will have to recreate the database.
 
-For the second form, first create your SRV record and publish it in DNS. This
-needs to be named _matrix._tcp.YOURDOMAIN, and point at at least one hostname
-and port where the server is running.  (At the current time synapse does not
-support clustering multiple servers into a single logical homeserver).  The DNS
-record would then look something like::
+If all goes well, you should be able to connect to your server with a client,
```
Comment (ara4n) [inline README.rst]: how? should probably link back to the "how to connect a client to your HS" section

File: README.rst [hunk 9]
```diff
@@ -454,140 +478,238 @@ versions of synapse.
...
+So, things to check are:
+
+* If you are trying to use a reverse-proxy, read `Reverse-proxying the
+  federation port`_.
+* If you are not using a SRV record, check that your ``server_name`` (the part
+  of your user-id after the ``:``) matches your hostname, and that port 8448 on
+  that hostname is reachable from outside your network.
+* If you *are* using a SRV record, check that it matches your ``server_name``
+  (it should be ``_matrix._tcp.<server_name>``), and that the port and hostname
+  it specifies are reachable from outside your network.
```
Comment (ara4n) [inline README.rst]: shouldn't we point people at the federation testing HTTP API?

File: README.rst [hunk 10]
```diff
@@ -454,140 +478,238 @@ versions of synapse.
...
+doing so is that it means that you can expose the default https port (443) to
+Matrix clients without needing to run Synapse with root privileges.
 
-If this is the first time you have used the client from that browser (it uses
-HTML5 local storage to remember its config), you will need to log in to your
-account. If you don't yet have an account, because you've just started the
-homeserver for the first time, then you'll need to register one.
+The most important thing to know here is that Matrix clients and other Matrix
+servers do not necessarily need to connect to your server via the same
+port. Indeed, clients will use port 443 by default, whereas other servers
```
Comment (ara4n) [inline README.rst]: s/other//

File: README.rst [hunk 11]
```diff
@@ -454,140 +478,238 @@ versions of synapse.
...
+Reverse-proxying the federation port
+------------------------------------
+
+There are two issues to consider before using a reverse-proxy on the federation
+port:
+
+* Due to the way SSL certificates are managed in the Matrix federation protocol
+  (see `spec <https://matrix.org/docs/spec/server_server/unstable.html#retrieving-server-keys>`_),
+  Synapse needs to be configured with the path to the SSL certificate, *even if
+  you do not terminate SSL at Synapse*.
```
Comment (ara4n) [inline README.rst]: I do not think this is true any more, since Mark's TLS changes - all synapse needs is to be told the fingerprint of the certificate, which can be passed in as a `tls_certificate_fingerprint` option or something like that?

File: README.rst [hunk 12]
```diff
@@ -151,38 +142,65 @@ This installs synapse, along with the libraries it uses, into a virtual
...
 change your Home Server's keys, you may find that other Home Servers have the
 old key cached. If you update the signing key, you should change the name of the
-key in the <server name>.signing.key file (the second word) to something different.
+key in the ``<server name>.signing.key`` file (the second word) to something different.
+
+The default configuration exposes two TCP ports: 8008 and 8448. Port 8008 is
+configured without TLS; it is not recommended this be exposed outside your
+local network. Port 8448 is configured to use TLS with a self-signed
+certificate. This is fine for testing with but you will almost certainly want
+to use another certificate for production purposes. You can do so by changing
```
Comment (ara4n) [inline README.rst]: this contradicts the doc later on, where we say that self-signed are fine.  i think we should spell out better when we can and can't get away with self-signed (i.e. client v. fed, again)

File: README.rst (hunk 9 above)
Comment (ara4n) [inline README.rst]: or perhaps provide a link to the bit where we mention it later on

File: README.rst [hunk 13]
```diff
@@ -454,140 +478,238 @@ versions of synapse.
...
 ================
 
-The job of authenticating 3PIDs and tracking which 3PIDs are associated with a
-given Matrix user is very security-sensitive, as there is obvious risk of spam
-if it is too easy to sign up for Matrix accounts or harvest 3PID data.
-Meanwhile the job of publishing the end-to-end encryption public keys for
-Matrix users is also very security-sensitive for similar reasons.
+Identity servers have the job of mapping email addresses and other 3rd Party
+IDs (3PIDs) to Matrix user IDs, as well as verifying the ownership of 3PIDs
+before creating that mapping.
```
Comment (ara4n) [inline README.rst]: in bold: **They are not where accounts or credentials are stored - these live on home serves.  Identity Servers are *just* for mapping 3rd party IDs to matrix IDs**.

Comment (ara4n) [inline README.rst]: (i know you say this at the end, but i think it bears reiterating)

File: README.rst [hunk 14]
```diff
@@ -601,24 +723,50 @@ server, they can request a password-reset token via clients such as Vector.
...
 
     UPDATE users SET password_hash='$2a$12$xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
         WHERE name='@test:test.com';
 
-Where's the spec?!
-==================
 
-The source of the matrix spec lives at https://github.com/matrix-org/matrix-doc.
-A recent HTML snapshot of this lives at http://matrix.org/docs/spec
+Synapse Development
```
Comment (ara4n) [inline README.rst]: I'm unsure about bumping synapse development so low here.  I almost feel we should give users the choice of either installing via pip or git (or OS package) from the outset, as it's common for people who install their own server to then want to fiddle with it a bit, and obviously a git checkout is the best platform for doing so...

File: README.rst [hunk 15]
```diff
@@ -601,24 +723,50 @@ server, they can request a password-reset token via clients such as Vector.
...
+    Confirm password:
     $2a$12$xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
 
-Then update the `users` table in the database:
+Then update the `users` table in the database::
 
     UPDATE users SET password_hash='$2a$12$xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
         WHERE name='@test:test.com';
 
-Where's the spec?!
```
Comment (ara4n) [inline README.rst]: Why delete the link to the spec?  This is kinda important...

File: README.rst (hunk 1 above)
Comment (richvdh) [inline README.rst]: well, where does it end? do we include the python sdk? the react sdk? I'll add a link to the "SDKs section" of the project page, which I think will be clearer than listing the SDKs here.

File: README.rst (hunk 2 above)
Comment (richvdh) [inline README.rst]: yup, have linked to the matrix.org blog post.

File: README.rst (hunk 15 above)
Comment (richvdh) [inline README.rst]: From the commit comment:

> - remove redundant "where's the spec" section: this would belong in "About matrix", but it's already there.

It could probably bear being made clearer in "about matrix", but I'm not sure that the synapse readme needs to be a comprehensive introduction to Matrix - that job is better done elsewhere on matrix.org.

File: README.rst (hunk 3 above)
Comment (richvdh) [inline README.rst]: Yeah, this is a difficult one, imho. The problem with trying to get people to use ``domain.com`` rather than ``mytestserver.domain.com`` from the get-go is that it means they need to grok the whole SRV business from the get-go, which is probably rather a lot to grasp.

As a compromise, I'll warn here that a server name can't be changed.

File: README.rst (hunk 5 above)
Comment (richvdh) [inline README.rst]: I've added a link to  the spec, but afaict this stuff is not well-documented.

File: README.rst (hunk 6 above)
Comment (richvdh) [inline README.rst]: Added some words.

File: README.rst (hunk 12 above)
Comment (richvdh) [inline README.rst]: added some words.

File: README.rst [hunk 16]
```diff
@@ -151,38 +142,65 @@ This installs synapse, along with the libraries it uses, into a virtual
...
-registration in the config by specifying ``enable_registration: true``
-(it is then recommended to also set up CAPTCHA - see docs/CAPTCHA_SETUP), or
-you can use the command line to register new users::
+You will need at least one user on your server in order to use a Matrix
+client. Users can be registered either `via a Matrix client`__, or via a
+commandline script.
+
+.. __: `client-user-reg`_
+
+To get started, is easiest to use the command line to register new users::
```
Comment (richvdh) [inline README.rst]: thanks, done

File: README.rst (hunk 7 above)
Comment (richvdh) [inline README.rst]: done

File: README.rst [hunk 17]
```diff
@@ -205,11 +234,54 @@ run (e.g. ``~/.synapse``), and::
...
+recommended to also set up CAPTCHA - see `<docs/CAPTCHA_SETUP.rst>`_.)
+
+Once ``enable_registration`` is set to ``true``, it is possible to register a
+user via `riot.im <https://riot.im/app/#/register>`_ or other Matrix clients.
+
+Your new user name will be formed partly from the ``server_name`` (see
+`Configuring synapse`_), and partly from a localpart you specify when you
+create the account. Your name will take the form of::
+
+    @localpart:my.domain.here
```
Comment (richvdh) [inline README.rst]: done, thanks

File: README.rst [hunk 18]
```diff
@@ -454,140 +478,238 @@ versions of synapse.
...
+yours to send messages.
+
+As explained in `Configuring synapse`_, the ``server_name`` in your
+``homeserver.yaml`` file determines the way that other servers will reach
+yours. By default, they will treat it as a hostname and try to connect to
+port 8448. This is easy to set up and will work with the default configuration,
+provided you set the ``server_name`` to match your machine's public DNS
+hostname.
+
+For a more flexible conversation, you can set up a DNS SRV record. This allows
```
Comment (richvdh) [inline README.rst]: thanks!

File: README.rst (hunk 10 above)
Comment (richvdh) [inline README.rst]: maybe. done.

File: README.rst (hunk 4 above)
Comment (richvdh) [inline README.rst]: I've tried to reword this a bit.

File: README.rst (hunk 13 above)
Comment (richvdh) [inline README.rst]: added

File: README.rst (hunk 14 above)
Comment (richvdh) [inline README.rst]: I disagree. The README is hard to grok as it is. I want to make it easier for people to get set up without support from us; anyone who gets that far and starts thinking about hacking on it will be able to find this.

File: README.rst (hunk 11 above)
Comment (richvdh) [inline README.rst]: well, maybe, but I know nothing about it. Can we make Mark document it here as a separate PR?

File: README.rst (hunk 14 above)
Comment (kyrias) [inline README.rst]: I think it might be worth it to split all the install instructions into an INSTALL file to get the README a bit more manageable.

**Comment by ara4n** [comment]
lgtm

File: README.rst (hunk 14 above)
Comment (richvdh) [inline README.rst]: @kyrias: that's a good suggestion, but it's hard to see where to draw the line. At the end of the day, most of the README is installation instructions, so this might just lead to a massive INSTALL file. (plus we tend to link to bits of it from everywhere).

Another option might be to split more bits of it out into the `docs` directory and link to them from the README.

Either way, I'd like to get this merged rather than blocking it until I have time to spend another day on it.
